### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by DenysMaksymKachalup**

### CODE REVIEW

File: src/main/java/com/example/springonlinebookstore/controller/AuthenticationController.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+@RestController
+@RequestMapping(value = "/auth")
+@RequiredArgsConstructor
+public class AuthenticationController {
+    private final UserService userService;
+
+    @PostMapping("/register")
+    public UserDto register(@RequestBody
+                                @Valid UserRegistrationRequestDto userRegistrationRequestDto)
+            throws RegistrationException {
```
Comment (sokimaaa) [inline src/main/java/com/example/springonlinebookstore/controller/AuthenticationController.java]:
```suggestion
    public UserDto register(@RequestBody
                                @Valid UserRegistrationRequestDto userRegistrationRequestDto)
            {
```

File: src/main/java/com/example/springonlinebookstore/controller/BookController.java [hunk 2]
```diff
@@ -43,7 +43,7 @@ public List<BookDto> findAll(Pageable pageable) {
     @GetMapping("/search")
     @Operation(summary = "Find book by parameters",
             description = "Get a list all books that match the parameters")
-    public List<BookDto> search(@RequestBody @Valid BookSearchParameters bookSearchParameters,
+    public List<BookDto> search(@RequestBody BookSearchParameters bookSearchParameters,
```
Comment (sokimaaa) [inline src/main/java/com/example/springonlinebookstore/controller/BookController.java]: Why remove valid? 
```suggestion
    public List<BookDto> search(@RequestBody @Valid BookSearchParameters bookSearchParameters,
```

File: src/main/java/com/example/springonlinebookstore/dto/user/UserRegistrationRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import org.hibernate.validator.constraints.Length;
+
+@Setter
+@Getter
+@EqualsAndHashCode
+@RequiredArgsConstructor
+@FieldsValueMatch.List({
+        @FieldsValueMatch(field = "password",
+                fieldMatch = "repeatPassword",
+                message = "The password fields must match")})
```
Comment (sokimaaa) [inline src/main/java/com/example/springonlinebookstore/dto/user/UserRegistrationRequestDto.java]: Why so complicate usage of annotation? make something like this:
```suggestion
@FieldsValueMatch(
        field = "password",
                fieldMatch = "repeatPassword",
                message = "The password fields must match")
                )
```

File: src/main/java/com/example/springonlinebookstore/exception/RegistrationException.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.example.springonlinebookstore.exception;
+
+public class RegistrationException extends Exception {
```
Comment (sokimaaa) [inline src/main/java/com/example/springonlinebookstore/exception/RegistrationException.java]: make it unchecked exception

File: src/main/java/com/example/springonlinebookstore/model/User.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import lombok.ToString;
+
+@Entity
+@Table(name = "users")
+@RequiredArgsConstructor
+@Getter
+@Setter
+@ToString
```
Comment (sokimaaa) [inline src/main/java/com/example/springonlinebookstore/model/User.java]:
```suggestion
```

File: src/main/java/com/example/springonlinebookstore/service/UserService.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package com.example.springonlinebookstore.service;
+
+import com.example.springonlinebookstore.dto.user.UserDto;
+import com.example.springonlinebookstore.dto.user.UserRegistrationRequestDto;
+import com.example.springonlinebookstore.exception.RegistrationException;
+
+public interface UserService {
+    UserDto register(UserRegistrationRequestDto userRegistrationRequestDto)
+            throws RegistrationException;
```
Comment (sokimaaa) [inline src/main/java/com/example/springonlinebookstore/service/UserService.java:9]:
```suggestion
```

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 7]
```diff
@@ -1,3 +1,5 @@
 databaseChangeLog:
   - include:
       file: db/changelog/changes/01-create-book-table.yaml
+  - include:
+      file: db/changelog/changes/02-create-user-table.yml
```
Comment (sokimaaa) [inline src/main/resources/db/changelog/db.changelog-master.yaml:5]: consider adding `03-insert-users`

File: src/main/java/com/example/springonlinebookstore/dto/user/UserRegistrationRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+@Setter
+@Getter
+@EqualsAndHashCode
+@RequiredArgsConstructor
+@FieldsValueMatch(field = "password",
+        fieldMatch = "repeatPassword",
+        message = "The password fields must match")
+public class UserRegistrationRequestDto {
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/springonlinebookstore/dto/user/UserRegistrationRequestDto.java]:
```suggestion
```

File: src/main/java/com/example/springonlinebookstore/model/User.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(unique = true, nullable = false)
+    private String email;
+    @Column(nullable = false)
+    private String password;
+    @Column(unique = true, nullable = false)
+    private String firstName;
+    @Column(unique = true, nullable = false)
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/springonlinebookstore/model/User.java]:
```suggestion
```

File: src/main/java/com/example/springonlinebookstore/model/User.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    private String lastName;
+    @Column(unique = true, nullable = false)
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/springonlinebookstore/model/User.java]:
```suggestion
```

File: src/main/java/com/example/springonlinebookstore/model/User.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+
+@Entity
+@Table(name = "users")
+@RequiredArgsConstructor
+@Getter
+@Setter
+public class User {
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/springonlinebookstore/model/User.java]:
```suggestion
```

File: src/main/java/com/example/springonlinebookstore/annotation/FieldsValueMatch.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    Class<?>[] groups() default {};
+    Class<? extends Payload>[] payload() default {};
+    String field();
+    String fieldMatch();
+    @Target({ElementType.TYPE, ElementType.ANNOTATION_TYPE})
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface List {
+        FieldsValueMatch[] value();
+    }
+
```
Comment (Sanderfuck) [inline src/main/java/com/example/springonlinebookstore/annotation/FieldsValueMatch.java]:
```suggestion
```

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

**Pull request merged**